### DISCUSSION TITLE
Dispatch custom event to RHF's `onChange` from `register`

### DISCUSSION THREAD

**Discussion by hmaesta** [react-hook-form/react-hook-form#12162] [Q&A] [answered]
I have a custom component that works as expected when wrapped into `Controller`. Still, I want to make things a bit less verbose and use this component with `register`, just like I do with text inputs.

<table>
<tr>
<td>Current</td>
<td>Desired</td>
</tr>
<tr>
<td>
<pre lang="tsx">
<‎Controller
   control={control}
   name="sport"
   render={({ field: { value, onChange } }) => (
      <‎MyComponent
         value={value}
         onChange={(newValue: string) => onChange(newValue)}
      />
   )}
/>
</pre>
</td>
<td>
<pre lang="tsx">
<‎MyComponent ㅤㅤㅤㅤㅤㅤㅤㅤㅤㅤㅤㅤㅤㅤㅤㅤㅤㅤ
   {...register('sport')}
/>
 ㅤ
ㅤ
ㅤ
ㅤ
ㅤ
ㅤ
ㅤ
</pre>
</td>
</tr>
</table>

Internally, `MyComponent` has no _real_ inputs, so dispatching a _real_ onChange event is not trivial. 

I [learned](https://github.com/react-hook-form/react-hook-form/blob/ea0f3ed86457691f79987a703ae8d50b9e16e2ad/src/types/form.ts#L182-L185) that `register` expects `onChange` to be [ChangeHandler](https://github.com/react-hook-form/react-hook-form/blob/ea0f3ed86457691f79987a703ae8d50b9e16e2ad/src/types/form.ts#L97-L100), so I made my component return an object following ChangeHandler's format, which didn't work. 

Next, after watching how native inputs behave with `register`, I tried dispatching [new Event](https://developer.mozilla.org/en-US/docs/Web/API/Event/Event) and [new CustomEvent](https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent/CustomEvent), but no success either. With this approach, I always get `Cannot read properties of null (reading 'name')`.

```tsx
const event = new CustomEvent('change', { bubbles: true, cancelable: true, detail: { value });
onChange(event)
```

Is there a way to dispatch a manually-created event on `onChange` that works with React Hook Form? 

→ **[Codesandbox exploration](https://codesandbox.io/p/sandbox/react-hook-form-register-onchange-pjh4vw?file=%2Fsrc%2FApp.tsx%3A50%2C2)** _(updated with working solution)_

**Comment by tinganho** [comment, marked as answer]
Succeded in "tricking it" to accept an object value.

```ts
onChange({ target: { name, value: myObject } as any); // important to include "name"
```

Their `onChange` handler, which we forced to branching into `getEventValue`:

<img width="584" alt="image" src="https://github.com/user-attachments/assets/473c1591-25d6-4c5e-8661-afd8fd77abb4">

Which is defined as:

<img width="467" alt="image" src="https://github.com/user-attachments/assets/085ecb8e-5676-4582-a9cd-97bf982dbd7c">

We forced it to evaluate `event.target.value`

**Reply by hmaesta** [reply]
It works! 🎉  Thank you!

[Updated Codesandbox](https://codesandbox.io/p/sandbox/react-hook-form-register-onchange-pjh4vw?file=%2Fsrc%2FApp.tsx%3A76%2C5) (line 77)

**Reply by tinganho** [reply]
I noticed that defaultValues where note propogated. I just created a class(ValueContainer) with getter and setter and called ref(new ValueContainer()) to set the new values. 

```ts
class ValueContainer<T> {
     private _value: T;
     constructor(public name: string, public onDefaultValue: (value: T) => void) // important that name is defined.
     set value(value: T) {
           this.onDefaultValue(value);
           this._value = value;
     }
}
```

And then in my component:
```ts
function MyComponent({ name, onChange, onBlur}, ref): JSX.Element {
     let defaultValue: any;
     ref(new ValueContainer(name, (v) => defaultValue = v));
}
```
